### ISSUE TITLE
enrich the .Mutable module apis in Vector

### ISSUE THREAD

**Issue by cartazio** [haskell/vector#203]
they're pretty poor in what they allow you to currently do. I thought i had a long standing ticket about this, but i cant seem to find it at the moment

@andrewthad @treeowl if you could bring some of the lovely energy to bear on the mutable subsystem of Vector, theres a HUGE amount of ways we can experiment with  improving things there, and that in turn can help motivate/justify ways to evolve primitive.

i've a lot of thoughts on this direction, but seldom the bandwidth to progress things as I'd like

**Comment by cartazio** [comment]
TODO: add links to the ticket discussions/ design notes around this if applicable

**Comment by andrewthad** [comment]
I'll try to get some ideas jotted down in the next week. I think the most difficult part is probably agreeing on an API. Since there's no stream-fusion to worry about in the mutable setting, implementing things should be straightforward.

**Comment by cartazio** [comment]
i'm more than happy to iterate on api 
sometimes i'm more responsive via irc (eg #numerical-haskell on freenode) and sometime email and sometime pestering on gh, but i think improving the mutable programming UX in vector is high impact and visible to more users than similar efforts in primitive (roughly speaking :) )

**Comment by cartazio** [comment]
current thoughts: have a `.Mutable.Generative`child module will have the "pure but on immutable" sibling of the mutable vector api

**Comment by konsumlamm** [comment]
I very much like the idea by @andrewthad, to add in-place `map`/`mapM`/`mapM_` functions. Some more ideas (using `MVector` from the `Data.Vector.Generic.Mutable` module):
* Versions of existing functions, but with arguments that operate in the given `PrimMonad`, for example:
  ```hs
  -- | Like 'modify', but with a monadic function.
  modifyM :: (PrimMonad m, MVector v a) => v (PrimState m) a -> (a -> m a) -> Int -> m ()
  ```
* In-place reverse
  ```hs
  -- | Reverse the vector in place.
  reverse :: (PrimMonad m, MVector v a) => v (PrimState m) a -> m ()
  ```

**Comment by Bodigrim** [comment]
> The functions `take`, `drop`, etc. imply that the argument should not be used again, although this is currently not documented. 

I would not say so. Documentation hints (by saying "without copying it") that slices are just a different view, but as long as you keep in mind that views are linked, it is perfectly valid to modify both input and output of `slice`; [`vector-algorithms`](http://hackage.haskell.org/package/vector-algorithms-0.8.0.3) do it all the time. This is not like `unsafeThaw`.

That said, I'd prefer `mapM :: (a -> m a) -> MVector s a -> m ()` to `mapM :: (a -> m a) -> MVector s a -> m (MVector s a)`.

> In-place reverse 

Side note: I use [`mreverse`](http://hackage.haskell.org/package/vector-rotcev-0.1.0.0/docs/Data-Vector-Rotcev.html#v:mreverse) for O(1) reversal of mutable vectors.

**Comment by Shimuuar** [comment]
Closed in favor of #334
